cli/mender-artifact/write.go

```diff
@@ -363,12 +364,34 @@ func applySoftwareVersionToTypeInfoProvides(ctx *cli.Context, typeInfoProvides a
 			typeInfoProvides = make(map[string]string)
 		}
 		for key, value := range softwareVersionMapping {
-			typeInfoProvides[key] = value
+			if typeInfoProvides[key] == "" || softwareVersionOverridesProvides(ctx, key) {
+				typeInfoProvides[key] = value
+			}
 		}
 	}
 	return typeInfoProvides
 }
 
+func softwareVersionOverridesProvides(ctx *cli.Context, key string) bool {
+	cmdLine := strings.Join(os.Args, " ")
+
+	var providesVersion string = `(-p|--provides)(\s+|=)` + regexp.QuoteMeta(key) + ":"
+	reProvidesVersion := regexp.MustCompile(providesVersion)
+	providesIndexes := reProvidesVersion.FindStringIndex(cmdLine)
```

Comment:
Should use `FindAllStringIndex` and return the last one, otherwise this won't work: `-p X --software-version Y -p Z`. This may seem overkill, but we expect many of the cases where these flags are used to be automated, so there may be a lot of redundant flags (imagine all the redundant flags on a gcc line after running through some Makefile logic).

Reply:
Done here: https://github.com/mendersoftware/mender-artifact/pull/291/commits/50ea75a6cd6e9e3d30311b24bfee258f32bc9209

To be honest, I could not write a test because the `-p X --software-version Y -p Z` flags trigger a `argument specified more than once` error and X is applied instead of Z

Reply:
That's weird, that's [being used in the tests](https://github.com/mendersoftware/mender-artifact/blob/master/cli/mender-artifact/dump_test.go#L165-L168).

Reply:
@kacf yes, but there you are using different keys, aren't you?

Reply:
Ah, good point. I think perhaps we should change this then. Being able to override the flags becomes more important with this new feature. Can you remove that message and change it to use the latest argument instead?

Reply:
Done here: https://github.com/mendersoftware/mender-artifact/pull/291/commits/98ef1e96df8359736b09a99d68cb7ab4c677d31e